compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -9065,7 +9074,26 @@ private STNode parseResourceAccessSegmentRhs(boolean isRhsExpr, boolean isInMatc
                 return parseResourceAccessSegmentRhs(isRhsExpr, isInMatchGuard);
         }
     }
-    
+
+    /**
+     * Parse on-fail-check node.
+     *
+     * <p>
+     * <code>
+     * on-fail-check := on-keyword fail-keyword variable-name => error-constructor
+     * </code>
+     *
+     * @return on-fail-check node
+     */
+    private STNode parseOnFailCheck() {
+        STNode onKeyword = parseOnKeyword();
+        STNode failKeyword = parseFailKeyword();
+        STNode varName = parseVariableName();
+        STNode rightDoubleArrow = parseDoubleRightArrow();
```

Comment:
Seems we call `recover(token, ParserRuleContext.EXPR_FUNC_BODY_START);` inside the `parseDoubleRightArrow()`. 
However, in this case the recovery context has to be `ParserRuleContext.RIGHT_DOUBLE_ARROW`.

Let's convert `parseDoubleRightArrow()` to `parseDoubleRightArrow(ParserRuleContext ctx)` and call,
`recover(token, ctx);`